internal/receiver/smartagentreceiver/filtering.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2021 Splunk, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smartagentreceiver
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/signalfx/golib/v3/datapoint"
+	"github.com/signalfx/signalfx-agent/pkg/core/config"
+	"github.com/signalfx/signalfx-agent/pkg/core/dpfilters"
+	"github.com/signalfx/signalfx-agent/pkg/monitors"
+	"github.com/signalfx/signalfx-agent/pkg/utils/filter"
+	"github.com/sirupsen/logrus"
+)
+
+type monitorFiltering struct {
+	filterSet       *dpfilters.FilterSet
+	metadata        *monitors.Metadata
+	hasExtraMetrics bool
+}
+
+func newMonitorFiltering(conf config.MonitorCustomConfig, metadata *monitors.Metadata) (*monitorFiltering, error) {
+	filterSet, err := buildFilterSet(metadata, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &monitorFiltering{
+		filterSet:       filterSet,
+		metadata:        metadata,
+		hasExtraMetrics: len(conf.MonitorConfigCore().ExtraMetrics) > 0 || len(conf.MonitorConfigCore().ExtraGroups) > 0,
+	}, nil
+}
+
+// AddDatapointExclusionFilter to the monitor's filter set.  Make sure you do this
+// before any datapoints are sent as it is not thread-safe with SendDatapoint.
+func (mf *monitorFiltering) AddDatapointExclusionFilter(filter dpfilters.DatapointFilter) {
+	mf.filterSet.ExcludeFilters = append(mf.filterSet.ExcludeFilters, filter)
+}
+
+func (mf *monitorFiltering) EnabledMetrics() []string {
+	if mf.metadata == nil {
+		return nil
+	}
+
+	// This can be optimized to cache results on the struct if it is worth it.
+	dp := &datapoint.Datapoint{}
+	var enabledMetrics []string
+
+	for metric := range mf.metadata.Metrics {
+		dp.Metric = metric
+		if !mf.filterSet.Matches(dp) {
+			enabledMetrics = append(enabledMetrics, metric)
+		}
+	}
+
+	return enabledMetrics
+}
+
+// HasEnabledMetricInGroup returns true if there are any metrics enabled that
+// fall into the given group.
+func (mf *monitorFiltering) HasEnabledMetricInGroup(group string) bool {
+	if mf.metadata == nil {
+		return false
+	}
+
+	for _, m := range mf.EnabledMetrics() {
+		// TODO: If metric names in metadata.yaml ever support wildcards this
```

Comment:
Is this TODO still valid for this project?

Reply:
removing